### ISSUE TITLE
Iteration does not stop if divisor is 1

### ISSUE THREAD

**Issue by Altreus** [rakudo/rakudo#4523]
https://github.com/rakudo/rakudo/blob/10b3272921e368545e9a02d02474a3760c85960c/src/core.c/Real.pm6#L74

The doc says "If the divisors are given as a lazy list, runs until the remainder is 0 or the list of divisors is exhausted."

But the code tests whether there is more to chop up, not whether the last remainder was zero. This, of course, remains constant if the divisor is 1, so you get a list of zeroes if you have an infinite list of ones.

```
m: say 10.polymod(1 xx *)
rakudo-moar 10b327292: OUTPUT: «(0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...)␤»
```

Both of these seem like reasonable behaviours, so I figured I'd open a discussion.

Roast does not test for this situation, or I'd defer to that.

(Incidentally, the doc also says "We can use lazy lists in polymod, as long as they are finite", but it doesn't seem to complain if it is infinite.)

**Comment by grondilu** [comment]
I don't get what "until the remainder is 0" means anyway.   Just because the remainder is zero doesn't mean there is nothing more to be done, otherwise this :

```
8.polymod(2 xx *)
```

would return (0) and not (0, 0, 0, 1)

**Comment by dakkar** [comment]
The docs and roast disagree. Roast says: https://github.com/Raku/roast/blob/master/S32-num/polymod.t#L11
```raku
is 1234567890.polymod(10 xx *), (0,9,8,7,6,5,4,3,2,1);
```
which clearly requires `polymod` to not stop when the remainder is 0 (as @grondilu pointed out, that would be a rather useless termination condition).

I suspect the docs meant to say "when the numerator inside the loop is 0" (but please use a better way of saying it).

**Comment by Altreus** [comment]
@grondilu You are correct, and it's not something our discussion had brought up until recently. Turns out you'd already noticed!

I suspect the doc means "until the *dividend* is zero", i.e. the number being - hopefully - reduced.

I can submit a doc bug if that's the consensus

**Comment by grondilu** [comment]
I think a recursive implementation, though probably not efficient, could help explain what this function is supposed to do exactly, thus clarifying the edge cases.

```
sub recursive_polymod($n, @divisors) {
  if @divisors {
    my $divisor = @divisors.shift;
    return $n < $divisor ?? $n !!
      flat ($n mod $divisor, recursive_polymod($n div $divisor, @divisors));
  } else {
    return $n
  }
}

say recursive_polymod(0x1234567890, my @ = 16 xx *);
say recursive_polymod(0x1234567890, my @ = 16 xx 5);

say 0x1234567890.polymod(16 xx *);
say 0x1234567890.polymod(16 xx 5);

say recursive_polymod(0, my @ = 16 xx *);
say recursive_polymod(0, my @ = 16 xx 5);
```

Notice that with that particular implementation, the case 1 xx * loops indefinitely, which kind of makes some sense IMHO.

**Comment by lizmat** [comment]
Tests added with https://github.com/Raku/roast/commit/bd0d3e068a

**Comment by lizmat** [comment]
https://gist.github.com/0784dbbe5e899b33e076fc449973d157

**Comment by lizmat** [comment]
Test fixed with https://github.com/Raku/roast/commit/2346529e63

**Comment by usev6** [comment]
I think that this can be closed again in the light of https://github.com/Raku/roast/commit/2346529e63.
